API Pull - Verify a nonce when setting Google WPCOM auth status

ianlin opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

This is a follow up from a PR comment: https://github.com/woocommerce/google-listings-and-ads/pull/2311#discussion_r1540827308.

The auth flow:

1. Merchant clicks enable new product sync
2. GL&A calls a new Google API to get the nonce, and add it in the `state` query parameter of the auth URL. Also save the nonce as an option in the DB.
3. GL&A redirects the merchant to the auth URL.
4. Merchant approves the access, then WPCOM will call Google’s callback URL (redirect URI as in the WPCOM OAuth document)
5. Google gets the nonce from the `state` query parameter, verify it, and get merchant’s WPCOM access token. Note that the “return URL” is also included in the `state` parameter.
6. Google redirects back to the merchant site, with query parameter `google_wpcom_app_status` and `nonce`.
7. GL&A verify the nonce parameter with the option in the DB, and save the `google_wpcom_app_status` status in the DB as well.

This PR implements the last step of the above flow, which it compares the nonce query parameter with the nonce option in the DB, and finally store the `google_wpcom_app_status` status in the DB.

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

**Prerequisite**
1. Make sure the notification is enabled by setting `add_filter( 'woocommerce_gla_notifications_enabled', '__return_true' )`.
2. Make sure the options `gla_wpcom_rest_api_status` and `gla_google_wpcom_auth_nonce` are not in the DB.

**Test Flow**
1. Complete the onboarding and go to the settings page.
2. Click `Enable` new product sync from the top-left banner.
3. It should redirect you to the WPCOM app granting access page.
4. Check the DB, the option `gla_google_wpcom_auth_nonce` should be set to `nonce-123`.
    - Note that since we are still waiting for the new Google API, the nonce is hardcoded as `nonce-123` at the moment.
5. Complete the auth flow, and go back to settings page with query parameters `google_wpcom_app_status=approved` and `nonce=nonce-123`, e.g.:
    - `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings&google_wpcom_app_status=approved&nonce=nonce-123`
6. The GMC card on the settings page should show `Google has been granted access to fetch your product data.`.
7. Check the DB, the option `gla_wpcom_rest_api_status` should be set as `approved`.
8. Check the DB, the option `gla_google_wpcom_auth_nonce` should be removed.
9. Redo the step 2 to 5, and use a wrong nonce query param other than `nonce-123` in the step 5.
10. Make sure the status is not being set, and you can see an error message in the browser's dev console
11. Redo the step 2 to 5, and use the correct nonce and another `google_wpcom_app_status` such as `disapproved` and `error`.
12. Make sure the status is set correctly in the DB and the UI is showing `There was an issue granting access to Google for fetching your products.`

### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Tweak - API Pull: Verify a nonce when setting Google WPCOM auth status

ianlin committed:
Save the nonce from Google's API when generating auth URL

ianlin committed:
Return option `google_wpcom_auth_nonce` in MC connected API

ianlin committed:
Verify nonces from URL query and the one stored in DB options

ianlin committed:
Move "maybe update auth status" from settings page to GMC accout card

As we need an option `google_wpcom_auth_nonce` that is stored in the
state `state.mc.accounts.mc.google_wpcom_auth_nonce`.

ianlin committed:
Delete the nonce after successfully update auth status in DB

ianlin committed:
Remove the nonce query param from the URL after auth status updated

To prevent the hook from trying to update the auth status again after
the connected-google-mc-account-card component gets updated.

ianlin committed:
Don't compare nonces if each of the nonce is not provided

Also don't remove nonce query param as it would actually still
re-render.

ianlin committed:
Also delete google wpcom auth nonce in DELETE /rest-api/authorize API

puntope commented on js/src/hooks/useUpdateRestAPIAuthorizeStatusByUrlQuery.js:
```diff
@@ -32,6 +37,16 @@ const useUpdateRestAPIAuthorizeStatusByUrlQuery = () => {
 
 	const handleUpdateRestAPIAuthorize = useCallback( async () => {
 		try {
+			if ( ! googleWPCOMAuthNonce || ! nonceFromURLQuery ) {
+				return;
+			}
+
+			if ( nonceFromURLQuery !== googleWPCOMAuthNonce ) {
+				throw new Error(
+					'Nonces mismatch when updating REST API authorize status.'
+				);
+			}
```
I wouldn't do the nonce verification in the Frontend as it might be altered on the client side. What about doing the check in the backend inside  `PUT rest-api/authorize`?  It seems more secure.

So we can send the query nonce and compare it in the backend with the saved in Options.

puntope reviewed (commented):
Thanks @ianlin for working on this. Please see my comment about avoiding the nonce check in the frontend.

ianlin committed:
Compare nonces in PUT /rest-api/authorize API

ianlin committed:
Remove nonces comparison in the frontend

ianlin committed:
Move 'verify nonce and update auth status' back to settings page

As we no longer require the state
`state.mc.accounts.mc.google_wpcom_auth_nonce` anymore.

ianlin committed:
Don't return `google_wpcom_auth_nonce` in API GET /mc/connection

ianlin committed:
Also delete google wpcom auth nonce when nonces comparison mismatch

ianlin committed:
Add more descriptions in method `update_wpcom_api_authorization()`

ianlin commented:
Thanks @puntope for reviewing and the nice suggestion, I should have been aware of doing nonce verification in the backend. I have updated it and it's ready for the 2nd round.

puntope commented on tests/Unit/MerchantCenter/AccountServiceTest.php:
```diff
@@ -770,4 +768,101 @@ public function test_disconnect() {
+		$nonce  = 'nonce-123';
+
+		$this->options->expects( $this->once() )
+			->method( 'get' )
```
💅  Maybe we can add the `with` method also here

puntope commented on js/src/hooks/useUpdateRestAPIAuthorizeStatusByUrlQuery.js:
```diff
@@ -21,7 +21,10 @@ import useApiFetchCallback from '.~/hooks/useApiFetchCallback';
-	const { google_wpcom_app_status: googleWPCOMAppStatus } = getQuery();
+	const {
+		google_wpcom_app_status: googleWPCOMAppStatus,
+		nonce: nonceFromURLQuery,
```
💅  Since now we have only one nonce maybe we can just keep this as `nonce`. Instead of redeclaring it as `nonceFromURLQuery`

puntope commented on src/MerchantCenter/AccountService.php:
```diff
@@ -508,12 +508,44 @@ public function reset_wpcom_api_authorization(): bool {
 
 	/**
 	 * Update the status of the merchant granting access to Google's WPCOM app in the database.
+	 * Before update status in the DB it will comapre the nonce stored in the DB and the nonce passed to the API.
```
Some typos here, "comapre" and some other grammar:

`Before updating the status in the DB it will compare the nonce stored in the DB with the nonce passed to the API.`

puntope commented on tests/Unit/MerchantCenter/AccountServiceTest.php:
```diff
@@ -770,4 +768,101 @@ public function test_disconnect() {
+			->method( 'delete' );
+
+		try {
+			$this->account->update_wpcom_api_authorization( $status, $nonce );
```
If `$this->account->update_wpcom_api_authorization` pass, then no assertion will happen and the test will pass. 

I saw in the code the we use

`$this->expectException` and `$this->expectExceptionMessage` for avoiding try catch

puntope commented on src/API/Site/Controllers/RestAPI/AuthController.php:
```diff
@@ -118,6 +118,7 @@ protected function delete_authorize_callback(): callable {
 		return function ( Request $request ) {
 			try {
 				$this->account_service->reset_wpcom_api_authorization();
```
I'm thinking if we should call the `delete_wpcom_api_auth_nonce` inside the `reset_wpcom_api_authorization` and `update_wpcom_api_authorization` just to be sure that if we move or call those in other areas of the codebase we don't forget to clear the nonce.

puntope commented on src/MerchantCenter/AccountService.php:
```diff
@@ -508,12 +508,44 @@ public function reset_wpcom_api_authorization(): bool {
+	 * @throws ExceptionWithResponseData If the stored nonce / nonce from query param is not provided, or the nonces mismatch.
 	 */
-	public function update_wpcom_api_authorization( string $status ): bool {
+	public function update_wpcom_api_authorization( string $status, string $nonce ): bool {
```
💅  Signature returns bool but PHPDOCS declares Sttring as return param

puntope reviewed (commented):
Hello @ianlin Thanks for the adjustments. Now it works in a much safer matter. 

I left tho some comments to be addressed about code and tests.

ianlin committed:
Test passing args in AccountServiceTest

ianlin committed:
Rephrase PHPDoc description for update_wpcom_api_authorization method

ianlin committed:
Update return type in PHPDoc for update_wpcom_api_authorization method

ianlin committed:
Use expectException to avoid using try/catch in PHP unit test

ianlin committed:
Delete nonce in update / reset auth status service, rather than in the
controller level

ianlin committed:
Simplify variable name by replacing nonceFromURLQuery with nonce

ianlin commented:
Hey @puntope, thanks for taking another look and the nice suggestions. I've pushed some commits regarding your suggestion and it's ready for another round.

puntope reviewed (approved):
Thanks for PR and further changes @ianlin LGTM ✅
